distribution/client/src/lib/snapshotter.js

```diff
@@ -0,0 +1,127 @@
+/*
+* Used for snapshotting and rollbacks during upgrades.
+* See:
+*  https://github.com/jenkinsci/jep/tree/master/jep/302
+*/
+
+const { spawnSync } = require('child_process');
+const logger        = require('winston');
+const fs            = require('fs');
+
+const LOG_PREFIX = '[snapshotting]';
+// https://github.com/jenkinsci/jep/tree/master/jep/302#user-content-files-to-store
+const GITIGNORE_CONTENT = `
+/plugins/
+/updates/
+/secrets/master.key
+`;
+
+class Snapshotter {
+
+  constructor(app, options) {
+    this.options = options || {};
+  }
+
+  init(workingDirectory) {
+    if (!workingDirectory) {
+      throw new Error('workingDirectory parameter is required');
+    }
+    this.workingDirectory = workingDirectory;
+
+    if (!fs.existsSync(this.getDotGitDirectory())) {
+      this.git('init');
+      this.git('config', 'user.email', 'evergreen-client@jenkins.io');
+      this.git('config', 'user.name', '"Evergreen Client Snapshotting System (JEP 302)"');
+      this.git('commit','--allow-empty', '--message', 'Initial evergreen commit');
+
+      fs.writeFile(`${this.getDotGitDirectory()}/description`,
```

Comment:
Is this intentionally async? Just curious

Reply:
Yes, see the message below for the error processing. I just do not want to stop anything just for this in case this goes wrong.